tests/unit/memefficiency.tcl

```diff
@@ -145,601 +290,326 @@ run_solo {defrag} {
+        foreach t $tests {
+            lassign $t test_proc type overrides
+            if {$type == "standalone"} {
+                start_server [list tags $standalone_tags overrides $overrides] {
+                    $test_proc $type
                 }
-
-                # in extreme cases of stagnation, we see over 20m misses before the tests aborts with "defrag didn't stop",
-                # in normal cases we only see 100k misses out of 500k elements
-                assert {$misses < $elements}
             }
-            # verify the data isn't corrupted or changed
-            set newdigest [debug_digest]
-            assert {$digest eq $newdigest}
-            r save ;# saving an rdb iterates over all the data / pointers
-            r del biglist1 ;# coverage for quicklistBookmarksClear
-        } {1}
-
-        test "Active defrag edge case: $type" {
-            # there was an edge case in defrag where all the slabs of a certain bin are exact the same
-            # % utilization, with the exception of the current slab from which new allocations are made
-            # if the current slab is lower in utilization the defragger would have ended up in stagnation,
-            # kept running and not move any allocation.
-            # this test is more consistent on a fresh server with no history
-            start_server {tags {"defrag"} overrides {save ""}} {
-                r flushdb sync
-                r config resetstat
-                r config set activedefrag no
-                r config set active-defrag-max-scan-fields 1000
-                r config set active-defrag-threshold-lower 5
-                r config set active-defrag-cycle-min 65
-                r config set active-defrag-cycle-max 75
-                r config set active-defrag-ignore-bytes 1mb
-                r config set maxmemory 0
-                set expected_frag 1.3
-
-                r debug mallctl-str thread.tcache.flush VOID
-                # fill the first slab containing 32 regs of 640 bytes.
-                for {set j 0} {$j < 32} {incr j} {
-                    r setrange "_$j" 600 x
-                    r debug mallctl-str thread.tcache.flush VOID
-                }
-
-                # add a mass of keys with 600 bytes values, fill the bin of 640 bytes which has 32 regs per slab.
-                set rd [valkey_deferring_client]
-                set keys 640000
-                for {set j 0} {$j < $keys} {incr j} {
-                    $rd setrange $j 600 x
-                }
-                for {set j 0} {$j < $keys} {incr j} {
-                    $rd read ; # Discard replies
+            if {$type == "cluster"} {
+                start_cluster 1 0 [list tags $cluster_tags overrides $overrides] {
+                    # Note: `start_cluster` passes the code through to another level which requires us
+                    #  to do an uplevel here.  Otherwise `test_proc` isn't recognized.
+                    uplevel 1 {$test_proc $type}
```

Comment:
This looks like a workaround for an error in `start_cluster`.

A correct solution could be to pass the number of uplevels upward. I mean `start_cluster` would pass the number 2 to `start_multiple_servers` which passes this on to `start_server`, so it can use `uplevel 2 $code`. I guess we could add an optional argument in the end, after the 'code' argument, in `start_multiple_servers` and `start_server`. Does it make sense?

Reply:
I don't plan to address existing design issues with `start_cluster` - that's outside the scope of defrag testing.  Not sure where that will lead.  It's disappointing that it doesn't work the same as `start_server`.

Reply:
Your workaround relies on this wrong behavior though, making this fail if we fix start_cluster later... Nah.. Let's make it a follow-up.

I almost have a solution in mind so I might do it for breakfast one of these days.